instrumentation/servlet/src/main/java/brave/servlet/ServletRuntime.java

```diff
@@ -72,38 +72,48 @@ private static ServletRuntime findServletRuntime() {
     }
 
     @Override void handleAsync(HttpServerHandler<HttpServletRequest, HttpServletResponse> handler,
-        HttpServletRequest request, Span span) {
+        HttpServletRequest request, HttpServletResponse response, Span span) {
       if (span.isNoop()) return; // don't add overhead when we aren't httpTracing
-      request.getAsyncContext().addListener(new TracingAsyncListener(handler, span));
+      TracingAsyncListener listener = new TracingAsyncListener(handler, request, response, span);
+      request.getAsyncContext().addListener(listener);
```

Comment:
Instead of adding request and response to the listener itself, you can also do

```java
request.getAsyncContext().addListener(listener, request, response);
```

That way, `e.getSuppliedRequest()` does not return null.

The advantage of your current solution is, however, that you don't need to cast, as `e.getSuppliedRequest()` returns a `ServletRequest`.

Reply:
Couldn't you call

```java
eventAsyncContext.addListener(this, event.getSuppliedRequest(), event.getSuppliedResponse())
```

when re-adding?

Reply:
acb2ed62 done! thanks for the idea